delivery_ups_oca/models/delivery_carrier.py

```diff
@@ -1,22 +1,285 @@
+            "label": "%s/ship/v1807/shipments/labels" % base_url,
+            "cancel": "%s/ship/v1/shipments/cancel" % base_url,
+            "status": "%s/track/v1/details/" % base_url,
+        }
+        return urls[ups_type]
+
+    def _ups_raise_error(self, api_result):
+        """If UPS' API returned an error, raise it"""
+        errors = api_result.get("response", {}).get("errors")
+        if self.debug_logging and errors:
+            raise UserError(
+                _("Sending %s to UPS: %s")
+                % (
+                    self.name,
+                    "\n".join("%(code)s %(message)s" % error for error in errors),
+                )
+            )
+
+    def _ups_set_headers(self):
+        return {
+            "AccessLicenseNumber": self.ups_access_license,
+            "Username": self.ups_ws_username,
+            "Password": self.ups_ws_password,
+            "transactionSrc": "Odoo (%s)" % self.env.cr.dbname,
+            "Content-Type": "application/json",
+            "Accept": "application/json",
+        }
+
+    def _ups_request(
+        self, url, payload=None, method="post", query_parameters=None,
+    ):
+        """Send a request to one of the UPS endpoints"""
+        self.ensure_one()
+        headers = self._ups_set_headers()
+        status = getattr(requests, method)(
+            url, json=payload, headers=headers, params=query_parameters
+        ).json()
+        self._ups_raise_error(status)
+        return status
 
     def ups_send_shipping(self, pickings):
-        return pickings._ups_send()
+        return [self.ups_create_shipping(p) for p in pickings]
+
+    def ups_create_shipping(self, picking):
+        """Send packages of the picking to UPS
+        return a list of dicts {'exact_price': 'tracking_number':}
+        suitable for delivery.carrier#send_shipping"""
+        self.ensure_one()
+        carrier = picking.carrier_id
+        status = carrier._ups_request(
+            url=carrier._ups_set_url("shipping"),
+            payload=carrier._ups_prepare_create_shipping(picking),
+        )
+        ShipmentResults = status["ShipmentResponse"]["ShipmentResults"]
+        total_charges = ShipmentResults["ShipmentCharges"]["TotalCharges"]
+        price = float(total_charges["MonetaryValue"])
+        if total_charges["CurrencyCode"] != picking.company_id.currency_id.name:
+            price = picking.company_id.currency_id._convert(
+                price,
+                self.env["res.currency"].search(
+                    [("name", "=", total_charges["CurrencyCode"])]
+                ),
+                picking.company_id,
+                fields.Date.today(),
+            )
+        picking.carrier_tracking_ref = ShipmentResults["ShipmentIdentificationNumber"]
+        # Add message to add .pdf file and auto-attachment
+        ShippingLabel = ShipmentResults["PackageResults"]["ShippingLabel"]
+        message = _("Tracking number: %s" % picking.carrier_tracking_ref)
```

Comment:
Is this is a usual practice? There's already a field for putting this information, and that field can have `tracking=True`